### PR TITLE
mgr/cephadm: Add disk rescan feature to the orchestrator

### PR DESCRIPTION
**Pull request by pcuzner**
This PR provides the ability to probe a hosts HBAs to pick up new/changed
devices. It is most useful for dense servers that use external disk enclosures
to provide OSDs.

Fixes: https://tracker.ceph.com/issues/55576
Signed-off-by: Paul Cuzner <pcuzner@redhat.com>

## Checklist
- Tracker (select at least one)
  - [X] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [X] New feature (ticket optional)
  - [] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [X] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by anthonyeleven — APPROVED** [review]
docs lgtm, but the code itself is beyond my scope so this isn't an approval of code.

File: doc/cephadm/host-management.rst
Comment (mgfritch) [inline doc/cephadm/host-management.rst:185]:
```suggestion
    ceph orch device rescan <hostname> [--with-summary]
```

would it make more sense for this to be placed under a `device` subcommand?

File: src/pybind/mgr/orchestrator/_interface.py
Comment (mgfritch) [inline src/pybind/mgr/orchestrator/_interface.py:424]:
```suggestion
        """
        Issue a disk rescan on every each HBA.
        
        Some HBA's and external enclosures don't automatically register
        device insert with the kernel, so for these scenarios we need
        to manually recan the hosts HBAs
        
        :param hostname: (str) host name
        """
```

Comment (mgfritch) [reply src/pybind/mgr/orchestrator/_interface.py:424]: (nit) use the same for docstring each impl of this func?

File: src/cephadm/cephadm
Comment (mgfritch) [inline src/cephadm/cephadm:9442]: seems ok, but wondering if this really should be a ceph-volume command instead?

File: doc/cephadm/host-management.rst
Comment (pcuzner) [reply doc/cephadm/host-management.rst:185]: I didn't think so. I can see what your saying but to me the action is against the host not against a device - and thats the way I read the syntax structure of the orch commands.

File: src/cephadm/cephadm
Comment (pcuzner) [reply src/cephadm/cephadm:9442]: There's a couple of  reason I placed it here 
- running ceph-volume requires the container to be instantiated - that's time. 1.5-2s on the physical machines I tested with
- it doesn't need anything in ceph-volume
- its not managing or changing any of the devices
- it's not going to be an async thing - its going to rely on responsiveness since the admin will issue the command for a rescan ad-hoc. The longer it takes the longer we tie up the cli thread
- it provides a more responsive UX

File: doc/cephadm/host-management.rst
Comment (pcuzner) [reply doc/cephadm/host-management.rst:185]: The other thing to consider is the UI to CLI relationship. In the UI you would perform a rescan against the 'host object', so having the CLI do the same thing is more intuitive...at least to me!

File: src/cephadm/cephadm
Comment (guits) [reply src/cephadm/cephadm:9442]: > seems ok, but wondering if this really should be a ceph-volume command instead?

yes, it could be, but in my opinion it doesn't have to.
If we implement this in `ceph-volume`, users will be bound to run it from within a container which is definitely not convenient and not needed for such a command/operation.

**Comment by pcuzner** [comment]
removed DNM label

File: doc/cephadm/host-management.rst
Comment (anthonyeleven) [inline doc/cephadm/host-management.rst:191]: Suggest s/adapters//

**Review by anthonyeleven — CHANGES_REQUESTED** [review]

File: doc/cephadm/host-management.rst
Comment (pcuzner) [reply doc/cephadm/host-management.rst:191]: done

File: src/pybind/mgr/orchestrator/_interface.py
Comment (anthonyeleven) [reply src/pybind/mgr/orchestrator/_interface.py:424]: wrt mgfritch's suggestion:

s/HBA's/HBAs/
s/insert/insertions/

**Review by anthonyeleven — APPROVED** [review]

**Comment by anthonyeleven** [comment]
Requested changed approved, for docs only.  @mgfritch suggested a change a few weeks ago that is unresolved, though, and I left a comment with fixes to that suggestion.

File: src/pybind/mgr/orchestrator/_interface.py
Comment (pcuzner) [reply src/pybind/mgr/orchestrator/_interface.py]: done

**Comment by pcuzner** [comment]
Applied the requested changes and squashed the commits.

test run
<pre>
[ceph: root@ceph-test-local /]# ceph orch host rescan ceph-test-local
Ok
[ceph: root@ceph-test-local /]# ceph orch host rescan ceph-test-local --with-summary
Ok. 6 adapters detected: 6 rescanned, 0 skipped, 0 failed (1.87s)
</pre>

I saw a tox issue during test_services.py - but this change isn't related to that module?

File: src/pybind/mgr/cephadm/module.py
Comment (anthonyeleven) [inline src/pybind/mgr/cephadm/module.py:1802]: s/every//

File: src/pybind/mgr/cephadm/module.py
Comment (anthonyeleven) [inline src/pybind/mgr/cephadm/module.py:1806]: s/recan/rescan
s/HBA/HBAs/
s/insert/insertions/

File: src/pybind/mgr/orchestrator/_interface.py
Comment (anthonyeleven) [inline src/pybind/mgr/orchestrator/_interface.py:423]: s/every//

File: src/pybind/mgr/orchestrator/_interface.py
Comment (anthonyeleven) [inline src/pybind/mgr/orchestrator/_interface.py:426]: s/insert/insertions/

File: src/pybind/mgr/orchestrator/module.py
Comment (anthonyeleven) [inline src/pybind/mgr/orchestrator/module.py:463]: Perform

File: doc/cephadm/host-management.rst
Comment (anthonyeleven) [inline doc/cephadm/host-management.rst:185]: Nit: s/disk/drive/

File: src/pybind/mgr/cephadm/module.py
Comment (anthonyeleven) [inline src/pybind/mgr/cephadm/module.py:1812]: Nit: is it too late to use ‘drive’ instead of ‘disk’?

**Review by anthonyeleven — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/cephadm/module.py
Comment (pcuzner) [reply src/pybind/mgr/cephadm/module.py:1812]: From my perspective,yes. However, if this gets merged feel free to raise a subsequent PR

File: src/pybind/mgr/orchestrator/module.py
Comment (pcuzner) [reply src/pybind/mgr/orchestrator/module.py:463]: done

File: src/pybind/mgr/orchestrator/_interface.py
Comment (pcuzner) [reply src/pybind/mgr/orchestrator/_interface.py:426]: done

File: src/pybind/mgr/orchestrator/_interface.py
Comment (pcuzner) [reply src/pybind/mgr/orchestrator/_interface.py:423]: done

File: src/pybind/mgr/cephadm/module.py
Comment (pcuzner) [reply src/pybind/mgr/cephadm/module.py:1806]: done

File: src/pybind/mgr/cephadm/module.py
Comment (pcuzner) [reply src/pybind/mgr/cephadm/module.py:1802]: done

File: doc/cephadm/host-management.rst
Comment (pcuzner) [reply doc/cephadm/host-management.rst:185]: reworded

**Review by anthonyeleven — APPROVED** [review]

**Comment by anthonyeleven** [comment]
Docs LGTM, should have code approval from someone else before merge.

**Comment by adk3798** [comment]
jenkins test make check

File: src/cephadm/cephadm
Comment (adk3798) [inline src/cephadm/cephadm:7997]: do we not actually care what `proc_name` is as long as it it isn't usb-storage or unknown?

File: src/cephadm/cephadm
Comment (adk3798) [inline src/cephadm/cephadm:8022]: for my clarification, this rescan isn't for actually figuring out what the devices are, it just triggers something that will allow us to find the devices later (with ceph-volume, I guess)

**Review by adk3798 — APPROVED** [review]
Code generally looks good to me. If I'm being honest I don't really understand what the rescan actually accomplishes (and I think a few of my review comments here reflect that).

File: src/cephadm/cephadm
Comment (pcuzner) [reply src/cephadm/cephadm:8022]: Correct. it's not telling you anything - it's probing the HBA(s) so the kernel can see new devices. On hardware that needs this, ceph-volume's next invocation will discover the new drives.

File: src/cephadm/cephadm
Comment (pcuzner) [reply src/cephadm/cephadm:7997]: unknown and usb-storage are two hba types that can cause problems - hang on probe. this is why we ignore

File: src/cephadm/cephadm
Comment (pcuzner) [reply src/cephadm/cephadm:7988]: Yes it's only used by rescan - however, I wrote it this way so if ever you need to run parallel tasks in cephadm you could reuse these functions.

File: src/pybind/mgr/rook/module.py
Comment (pcuzner) [reply src/pybind/mgr/rook/module.py:436]: ACK - do you want me to remove?

**Comment by pcuzner** [comment]
@adk3798 the reason for rescan stems from dealing with external enclosures. Working on the big WDC enclosures device removal and insert did not get passed to the kernel automatically, and it needed a rescan to pick up the changes. It's the same deal if we were picking up devices from a SAN - in those environments rescanning to pick up changes is the norm.

File: src/pybind/mgr/rook/module.py
Comment (anthonyeleven) [reply src/pybind/mgr/rook/module.py:436]: If not, I suggest adding an "s" to "recan"

**Comment by adk3798** [comment]
jenkins test api

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2022-07-26_01:35:13-orch:cephadm-wip-adk-testing-2022-07-25-1649-distro-default-smithi/

lots of rhel tests failing due to an issue with satellite, reruns

https://pulpito.ceph.com/adking-2022-07-26_14:22:07-orch:cephadm-wip-adk-testing-2022-07-25-1649-distro-default-smithi/

1 actual failure was known issue with agent/on test_nfs timing out

**Pull request merged**
